### PR TITLE
Add support for XEP-0221: Data Forms Media Element.

### PR DESCRIPTION
**Pull request by adiaholic**
While I was working on XEP-0232 (Software Information),
I felt the need of XEP-0221 (Data Forms Media Element).
So I gave it a try.

As you will go through the code, you will realize that
I struggled a bit with these issues :

`URINode.java` :
I think that it's constructor impementation is a bit unconventional.
Inside the toXML funtion,
xml.attribute doesn't work because it gets confused for `APOS_ENCODE`,
so I harcoded it for now.

Also this work is still in progress but any suggestions are welcome :)

### CODE REVIEW

File: smack-extensions/src/main/java/org/jivesoftware/smackx/dataformmedia/MediaElement.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        if (height != null && width != null) {
+            xml.attribute(attribute_height, Integer.toString(height));
+            xml.attribute(attribute_width, Integer.toString(width));
+        }
+        xml.rightAngleBracket();
+        xml.append(getMediaXML());
+        return xml.closeElement(getElementName());
+    }
+
+    private CharSequence getMediaXML() {
```
Comment (vanitasvitae) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/dataformmedia/MediaElement.java]: You should probably return XmlStringBuilder here.

File: smack-extensions/src/main/java/org/jivesoftware/smackx/dataformmedia/MediaElement.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            mediaXML.element(node);
+        }
+        return mediaXML;
+    }
+
+    public static Builder getBuilder() {
+        return new Builder();
+    }
+
+    public static final class Builder{
```
Comment (vanitasvitae) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/dataformmedia/MediaElement.java]: Not sure if the Builder pattern is useful here, as `build(...)` is the only method. Making the constructors public would probably be sufficient in this situation.

File: smack-extensions/src/main/java/org/jivesoftware/smackx/dataformmedia/MediaType.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jivesoftware.smackx.dataformmedia;
+
+public class MediaType {
+    public static String audio_xwav = "audio/x-wav";
```
Comment (vanitasvitae) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/dataformmedia/MediaType.java]: Those Strings should probably be final.

File: smack-extensions/src/main/java/org/jivesoftware/smackx/dataformmedia/URINode.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    private final String ELEMENT = "uri";
+    private final String type_subtype;
+
+    public URINode(String type, URI uri) {
+        this.type_subtype = "'" + type + "'";
+        this.uri = uri;
+    }
+
+    public URINode(String type, String subType, URI uri) {
+        this.type_subtype = "'" + type + "; " + subType + "'";
```
Comment (vanitasvitae) [inline smack-extensions/src/main/java/org/jivesoftware/smackx/dataformmedia/URINode.java]: You may want to do some null-checks here.

File: smack-extensions/src/test/java/org/jivesoftware/smackx/dataformmedia/MediaElementTest.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        URI uri1 = null;
+        URI uri2 = null;
+        URI uri3 = null;
+
+        try {
+            uri1 = new URI("http://victim.example.com/challenges/speech.wav?F3A6292C");
+            uri2 = new URI("cid:sha1+a15a505e360702b79c75a5f67773072ed392f52a@bob.xmpp.org");
+            uri3 = new URI("http://victim.example.com/challenges/speech.mp3?F3A6292C");
+        } catch (URISyntaxException e) {
+             throw new IllegalArgumentException("Invalid URI");
```
Comment (vanitasvitae) [inline smack-extensions/src/test/java/org/jivesoftware/smackx/dataformmedia/MediaElementTest.java]: I'd rather throw an `AssertionError` here, but that doesn't really matter too much.

Comment (Flowdalic) [reply smack-extensions/src/test/java/org/jivesoftware/smackx/dataformmedia/MediaElementTest.java]: Besides from swallowing `e` here, why would you catch the `URISyntaxException` anyways? Just declare that the method throws it.

File: smack-extensions/src/test/java/org/jivesoftware/smackx/dataformmedia/MediaElementTest.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+                   "cid:sha1+f24030b8d91d233bac14777be5ab531ca3b9f102@bob.xmpp.org" +
+                   "</uri>" +
+                   "</media>";
+           URI uri1 = null;
+           URI uri2 = null;
+           try {
+               uri1 = new URI("http://www.victim.com/challenges/ocr.jpeg?F3A6292C");
+               uri2 = new URI("cid:sha1+f24030b8d91d233bac14777be5ab531ca3b9f102@bob.xmpp.org");
+           } catch (URISyntaxException e) {
+                   throw new IllegalArgumentException("Invalid URI");
```
Comment (Flowdalic) [inline smack-extensions/src/test/java/org/jivesoftware/smackx/dataformmedia/MediaElementTest.java]: Same here

**Comment by adiaholic** [comment]
Thank you for your valuable suggestions.
What I have done until now is 
provided minimal support to satisfy XEP-0221 examples.
I wish to add support for more type/subtype through
[IANA MIME MEDIA TYPE REGISTRY](https://www.iana.org/assignments/media-types/media-types.xhtml)
I could manually add some more type/subtypes, ofcourse I have no issue with it.
But I'm afraid I'll make a fool out of myself.
:)

**Comment by vanitasvitae** [comment]
I'd simply let it up to the user to decide, which media types they want to put into the elements and what types they expect to receive. As far as I understand a registry is subject to constant change (new types are added / removed regularly), so I don't see much point in maintaining a class / enum with all the types.

You should however add the link to the type registry to the javadoc of the corresponding methods, so users can quickly find out what to put into the methods.

**Comment by adiaholic** [comment]
Seems fair to me !

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: eb99d802d34e25b49a8862847f746a854b2ce682
  * Your commit message body contains a line that is longer than 72 characters

Guidelines are available at https://github.com/igniterealtime/Smack/wiki/Guidelines-for-Smack-Developers-and-Contributors

---
This message was auto-generated by https://gitcop.com

**Comment by Flowdalic** [comment]
Closing in favor of https://github.com/igniterealtime/Smack/commit/4d36e3b5211dd65d2d8d31223b5c02109f309b03

**Pull request closed without merging**